PROD-689: Automated banned users

samott opened the pull request:
Pulls suspected bot users from the mechanism engine and updates the database.

Runs 10 minutes before every hour so questions that are about to reveal are included in the analysis.

The production endpoint is not deployed yet.

marvinmarnold commented on lib/jwt.ts:
```diff
@@ -5,7 +5,7 @@ import { cookies } from "next/headers";
   const user = await prisma.user.findFirst({ where: { id: userId } });
   if (!user) throw new Error("User not found");
-  if (user.threatLevel) {
+  if (user.threatLevel === "bot") {
```
We could make this a constant/enum.

marvinmarnold commented on lib/bots.ts:
```diff
@@ -0,0 +1,19 @@
+    },
+    where: {
+      id: { in: bots },
+      threatLevel: null,
```
I know I said we should prevent overwriting users who have been pardoned. I don't think we should. 

We just need to make sure that in the future if we pardon someone that we don't pardon them until after the window that mechanism engine looks at has passed. Otherwise the same user will be immediately re-identified as a bot.

Scope increase, nice to have:
Ideally, the case where a pardoned user is detected botting again could throw a Sentry event that we could then turn into a Slack alert with clear message about the username, address, and userId suspected of being a bot.

marvinmarnold commented:
I just merged the mechanism code for this fyi

samott commented:
@marvinmarnold I updated the code with more features. The code automatically checks if it's changed the threat level in the current analysis window (I updated the mechanism engine to output the dates used), and there are a few different enums at our disposal to control re-blocking behaviour.

marvinmarnold commented on app/api/cron/detect-bot-activity/route.ts:
```diff
@@ -0,0 +1,72 @@
+      },
+    );
+  } catch (error) {
+    console.error("Error:", error);
```
I believe error logs get picked up by Sentry but I can't remember. The most likely reason this will happen is if mechanism-engine runs out of memory. Can we do Sentry.captureMessage here instead and make it more explicit what the error is?

marvinmarnold reviewed (commented):
Code looks good. But some of the test failures look like they are related to this code.
